cmd/contour/serve.go

```diff
@@ -161,235 +163,202 @@ func registerServe(app *kingpin.Application) (*kingpin.CmdClause, *serveContext)
 	return serve, ctx
 }
 
-// doServe runs the contour serve subcommand.
-func doServe(log logrus.FieldLogger, ctx *serveContext) error {
+type Server struct {
+	group    workgroup.Group
+	log      logrus.FieldLogger
+	ctx      *serveContext
+	clients  *k8s.Clients
+	mgr      manager.Manager
+	registry *prometheus.Registry
+}
+
+func NewServe(log logrus.FieldLogger, ctx *serveContext) (*Server, error) {
+
+	// Set up workgroup runner.
+	var group workgroup.Group
+
 	// Establish k8s core & dynamic client connections.
 	clients, err := k8s.NewClients(ctx.Config.Kubeconfig, ctx.Config.InCluster)
 	if err != nil {
-		return fmt.Errorf("failed to create Kubernetes clients: %w", err)
+		return nil, fmt.Errorf("failed to create Kubernetes clients: %w", err)
```

Comment:
nit on what `NewServe` returns, seems like we have a mix of `log.Fatal` and returning an error, should we standardize on one?

Reply:
good catch! Copy pasted that all together. Updated to make it consistent.